### PR TITLE
[WIP][SERVER-2078] Initial air-gap documentation

### PR DESCRIPTION
**Pull request by circlery**
# Description
Draft pull request - needs more details added around copying images, nomad configuration, runner installation.

# Reasons
SERVER-2078

# Content Checklist
- [ ] Break up walls of text by adding paragraph breaks.
- [ ] Consider if the content could benefit from more structure, such as lists or tables, to make it easier to consume.
- [ ] Keep the title between 20 and 70 characters.
- [ ] Consider whether the content would benefit from more subsections (h2-h6 headings) to make it easier to consume.
- [ ] Check all headings h1-h6 are in sentence case (only first letter is capitalized).
- [ ] Is there a "Next steps" section at the end of the page giving the reader a clear path to what to read next?
- [ ] Include relevant backlinks to other CircleCI docs/pages.

[SERVER-2078]: https://circleci.atlassian.net/browse/SERVER-2078?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: jekyll/_cci2/server/air-gapped-installation/phase-2-configure-object-storage.adoc [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+:page-description: How to configure object storage through Min.IO to run CircleCI Server in an air-gapped environment.
+:icons: font
+:toc: macro
+:toc-title:
+
+[#configure-minio]
+== Initial Min.IO Setup
+
+[#create-airgap-data-bucket]
+=== 1. Create a airgap-data bucket.
```
Comment (lloydChris) [inline jekyll/_cci2/server/air-gapped-installation/phase-2-configure-object-storage.adoc:18]: This bucket is never referenced again, and I am not sure why the values.yaml would go there.  I think i can be removed

File: jekyll/_cci2/server/air-gapped-installation/phase-2-configure-object-storage.adoc [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+Create a new bucket in min.io named `airgap-data`. This bucket will be for your values.yaml file.
+
+[#create-circle-data-bucket]
+=== 2. Create a circle-data bucket.
+Create a new bucket in min.io named `circle-data`.
+
+[#configure-circle-data-bucket]
+=== 3. Configure the circle-data bucket.
```
Comment (lloydChris) [inline jekyll/_cci2/server/air-gapped-installation/phase-2-configure-object-storage.adoc:26]:
```suggestion
=== 2. Create a circleci-data bucket.
Create a new bucket in min.io named `circleci-data`.

[#configure-circle-data-bucket]
=== 3. Configure the circleci-data bucket.
```

File: jekyll/_cci2/server/air-gapped-installation/phase-3-installing-circleci-server.adoc [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+[source, yaml]
+----
+distributor:
+  agent_base_url: http://<minio-internal-tld>:9000/circle-data/
+  launch_agent_base_url: http://<minio-internal-tld>:9000/circle-data/
+----
+
+[#configure-minio-storage]
+=== 2. Configure Min.IO Object Storage
```
Comment (lloydChris) [inline jekyll/_cci2/server/air-gapped-installation/phase-3-installing-circleci-server.adoc:31]: This is covered in the regular docs: https://circleci.com/docs/server/installation/phase-2-core-services/#s3-compatible

File: jekyll/_cci2/server/air-gapped-installation/phase-3-installing-circleci-server.adoc [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+Specify the location of the Nomad build agent image within your registry, copied during phase 1, modifying the port as necessary.
+
+[source, yaml]
+----
+nomad:
+  buildAgentImage: "<your-internal-registry-tld>:5000/circleci/picard"
+----
+
+[#configure-docker-registry]
+=== 4. Configure your internal docker registry
```
Comment (lloydChris) [inline jekyll/_cci2/server/air-gapped-installation/phase-3-installing-circleci-server.adoc:57]: This is covered in the regular docs: https://circleci.com/docs/server/installation/phase-2-core-services/#Registry

File: jekyll/_cci2/server/air-gapped-installation/phase-3-installing-circleci-server.adoc [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+[source, yaml]
+----
+tls:
+  import:
+    - "<github-enterprise-internal-hostname>:443"
+----
+
+
+[#configure-ghe]
+=== 6. Configure GitHub Enterprise
```
Comment (lloydChris) [inline jekyll/_cci2/server/air-gapped-installation/phase-3-installing-circleci-server.adoc:81]: This is covered in the regular docs: https://circleci.com/docs/server/installation/phase-2-core-services/#github-enterprise-integration

File: jekyll/_cci2/server/air-gapped-installation/phase-4-configuring-nomad-clients.adoc [hunk 6]
```diff
@@ -0,0 +1,47 @@
+---
+version:
+- Server v4.x
+- Server Admin
+---
+= Phase 4 - Configuring Nomad Clients
```
Comment (lloydChris) [inline jekyll/_cci2/server/air-gapped-installation/phase-4-configuring-nomad-clients.adoc:6]: I don't know if we want to tell them how to install nomad.  I wonder if we should point them to the nomad docs instead.  Then we have less risk of falling out of date with their process and version numbers

**Review by lloydChris — CHANGES_REQUESTED** [review]
Good job on these, there is a lot of information to cover

we don't use the term "tutorial" to describe our install instructions anywhere else, so i don't think we want to introduce that here.

File: jekyll/_cci2/server/air-gapped-installation/overview.adoc [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+:page-liquid:
+:page-description: A tutorial on how to install CircleCI Server in an air-gapped environment.
+:icons: font
+:toc: macro
+:toc-title:
+
+
+CircleCI Server, while normally installed in an internet-connected cloud environment, can be installed in an air-gapped environment with a few prerequisites and additional steps.
+
+This tutorial will walk you through the prerequisites required, and the configuration changes needed in order to install CircleCI Server in an air-gapped environment.
```
[the remark this answers is not in the document]
Comment (circlery) [reply jekyll/_cci2/server/air-gapped-installation/overview.adoc:17]: Good call

File: jekyll/_cci2/server/air-gapped-installation/overview.adoc [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+[#components]
+== CircleCI Server - Air Gapped Components
+The following is an overview of the prerequisites required to run an air-gapped CircleCI Server installation. These items must be present in your air-gapped environment in order to move forwards with the installation.
+
+[.table.table-striped]
+[cols=4*, options="header", stripes=even]
+|===
+| Component
+| Location
```
[the remark this answers is not in the document]
Comment (circlery) [reply jekyll/_cci2/server/air-gapped-installation/overview.adoc:29]: Good call - Originally it was a table of everything required, including the stuff in the tutorial. I'll remove this.

File: jekyll/_cci2/server/air-gapped-installation/overview.adoc [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+| An object storage system (min.io)
+| Air-gapped environment
+| Object and artifact storage.
+| 
+
+| A version control system
+| Air-gapped environment
+| Source of code for CircleCI Server to build.
+| We recommend using GitHub enterprise.
```
[the remark this answers is not in the document]
Comment (circlery) [reply jekyll/_cci2/server/air-gapped-installation/overview.adoc:51]: Perfect!

File: jekyll/_cci2/server/air-gapped-installation/overview.adoc [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+| Location
+| Used for
+| Notes
+
+| Container registry
+| Air-gapped environment
+| Hosting CircleCI server images for Kubernetes.
+| 
+
+| A Kubernetes installation
```
[the remark this answers is not in the document]
Comment (circlery) [reply jekyll/_cci2/server/air-gapped-installation/overview.adoc:38]: 👍

File: jekyll/_cci2/server/air-gapped-installation/phase-1-copy-images.adoc [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+:page-layout: classic-docs
+:page-liquid:
+:page-description: How to copy the necessary images and charts for CircleCI Server to your air-gapped environment.
+:icons: font
+:toc: macro
+:toc-title:
+
+CircleCI Server is deployed into a Kubernetes cluster using a Helm chart. As such, we'll need to copy the latest images and charts from CircleCI into your virtual environment.
+
+[#copy-images]
```
[the remark this answers is not in the document]
Comment (circlery) [reply jekyll/_cci2/server/air-gapped-installation/phase-1-copy-images.adoc:16]: @lloydChris They'll need to authenticate / pull from our registry no matter what. I left the push/copy docs vague, but do we want to keep the pull stuff in here?

File: jekyll/_cci2/server/air-gapped-installation/phase-2-configure-object-storage.adoc (hunk 2 above)
Comment (circlery) [reply jekyll/_cci2/server/air-gapped-installation/phase-2-configure-object-storage.adoc:26]: Waiting on confirmation before changing this:
https://circleci.atlassian.net/wiki/spaces/server/pages/6650560666/Test+Case+-+Air+Gap

Comment (lloydChris) [reply jekyll/_cci2/server/air-gapped-installation/phase-2-configure-object-storage.adoc:26]: I am missing something, confirmation of what?

**Comment by rosieyohannan** [comment]
Hello @circlery! I'm planning to start a versioning system for the 4.1 release so these docs will end up in a different location. Can you let me know when the content is ready for me to start working on and I'll branch off this branch to get the pages in the correct location etc. 🙏

File: jekyll/_cci2/server/air-gapped-installation/phase-3-installing-circleci-server.adoc (hunk 3 above)
Comment (circlery) [reply jekyll/_cci2/server/air-gapped-installation/phase-3-installing-circleci-server.adoc:31]: I'd prefer we keep it here if possible. We're more explicit here in our definition of what's required in the yaml in the context of an air-gapped installation.

File: jekyll/_cci2/server/v4.1/air-gapped-installation/phase-2-configure-object-storage.adoc [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+- Server Admin
+---
+= Phase 2 - Configure object storage
+:page-layout: classic-docs
+:page-liquid:
+:page-description: How to configure object storage through MinIO to run CircleCI Server in an air-gapped environment.
+:icons: font
+:toc: macro
+:toc-title:
+
```
Comment (rosieyohannan) [inline jekyll/_cci2/server/v4.1/air-gapped-installation/phase-2-configure-object-storage.adoc:13]: @circlery @lloydChris I think we could do with a bit of context for this page - just an intro about what the object storage is used for, and why we use minIO. If you could give me some brief notes I can add something.

File: jekyll/_cci2/server/v4.1/air-gapped-installation/phase-5-configure-circleci-runner.adoc [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+- Server Admin
+---
+= Phase 5 - Configuring CircleCI Runner
+:page-layout: classic-docs
+:page-liquid:
+:page-description: How to configure CircleCI Runner to run with CircleCI Server in an air-gapped environment.
+:icons: font
+:toc: macro
+:toc-title:
+
```
Comment (rosieyohannan) [inline jekyll/_cci2/server/v4.1/air-gapped-installation/phase-5-configure-circleci-runner.adoc:13]: @circlery @lloydChris Same as above, some context about why we are installing runner at this point would be useful

File: jekyll/_cci2/server/v4.1/air-gapped-installation/phase-5-configure-circleci-runner.adoc [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+./circleci --skip-update-check namespace create <github-enterprise-username> github <github-enterprise-username>
+----
+
+[#create-resource-class]
+== 8. Create a resource class
+On the air-gapped machine, run the following command to create a resource class with your GitHub enterprise username:
+
+[source,shell]
+----
+./circleci --skip-update-check runner resource-class create <github-enterprise-username>/default "Default Runners" --generate-token
```
Comment (rosieyohannan) [inline jekyll/_cci2/server/v4.1/air-gapped-installation/phase-5-configure-circleci-runner.adoc:80]: @circlery in this command do they only need to sub their GHE username? Or is "Default Runners" something that will be customer-specific?

**Comment by rosieyohannan** [comment]
@circlery If all @lloydChris's change requests have been addressed could you mark them as resolved?

**Comment by rosieyohannan** [comment]
@circlery and @lloydChris I propose to merge these changes as they are into the release branch and I'll move over the questions for you in this PR too. If my questions don't make sense just let me know and I'll rephrase 😆 If we don't get time to address them I'll pull them into a Jira ticket to chase up after the release :-D

**Comment by rosieyohannan** [comment]
@circlery I will wait for a 👍 from you before merging this as it's your PR and still set to WIP

File: jekyll/_cci2/server/v4.1/air-gapped-installation/phase-2-configure-object-storage.adoc (hunk 12 above)
Comment (circlery) [reply jekyll/_cci2/server/v4.1/air-gapped-installation/phase-2-configure-object-storage.adoc:13]: added

**Pull request merged**